integrations/terraform_cloud/main.py

```diff
@@ -0,0 +1,144 @@
+from enum import StrEnum
+from typing import Any
+from client import TerraformClient
+from port_ocean.context.ocean import ocean
+from loguru import logger
+from asyncio import gather
+from port_ocean.core.ocean_types import ASYNC_GENERATOR_RESYNC_TYPE
+import asyncio
+
+
+class ObjectKind(StrEnum):
+    WORKSPACE = "workspace"
+    RUN = "run"
+    STATE_VERSION = "state-version"
+
+
+def init_terraform_client() -> TerraformClient:
+    """
+    Intialize Terraform Client
+    """
+    config = ocean.integration_config
+
+    terraform_client = TerraformClient(
+        config["terraform_host"],
+        config["terraform_token"],
+    )
+
+    return terraform_client
+
+
+## Enriches the state version with output
+async def enrich_state_versions_with_output_data(
+    http_client: TerraformClient, state_versions: list[dict[str, Any]]
+) -> list[dict[str, Any]]:
+    async with asyncio.BoundedSemaphore(5):
+        tasks = [
+            http_client.get_state_version_output(state_version["id"])
+            for state_version in state_versions
+        ]
+
+        output_batches = await asyncio.gather(*tasks)
+
+        enriched_state_versions = [
+            {**state_version, "__output": output}
+            for state_version, output in zip(state_versions, output_batches)
+        ]
+
+        return enriched_state_versions
+
+
+@ocean.on_resync(ObjectKind.WORKSPACE)
+async def resync_workspaces(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    terraform_client = init_terraform_client()
+    async for workspace in terraform_client.get_paginated_workspaces():
+        logger.info(f"Received {len(workspace)} batch {kind}")
+        yield workspace
+
+
+@ocean.on_resync(ObjectKind.RUN)
+async def resync_runs(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    terraform_client = init_terraform_client()
+
+    async def fetch_runs_for_workspace(workspace: dict[str, Any]) -> list[Any]:
+        return [
+            run
+            async for run in terraform_client.get_paginated_runs_for_workspace(
+                workspace["id"]
+            )
+        ]
+
+    async def fetch_runs_for_all_workspaces() -> ASYNC_GENERATOR_RESYNC_TYPE:
+        async for workspaces in terraform_client.get_paginated_workspaces():
+            logger.info(
+                f"Received {len(workspaces)} batch workspaces... fetching its associated {kind}"
+            )
+
+            tasks = [fetch_runs_for_workspace(workspace) for workspace in workspaces]
+            runs_batches = await asyncio.gather(*tasks)
+
+            for runs in runs_batches:
+                for run in runs:
+                    yield run
+
```

Comment:
Why one at a time?

Reply:
Here the integration in unable to upsert the entities to Port when it's in batches ([[some data]]) unless I yield it one by one

Reply:
Each run is a list?
It should be able to handle lists